### DISCUSSION TITLE
Libadwaita ?

### DISCUSSION THREAD

**Discussion by rene-coty** [pdfarranger/pdfarranger#775] [Q&A] [answered]
I like this application very much; it would be great to have a port to GTK4 & Libadwaita, therefore the UX would be more polished and the application would better integrate in the modern desktop environments..

**Comment by jeromerobert** [comment, marked as answer]
I started to test PDF Arranger with GTK4 but there are still lot of work to make it work (see https://github.com/jeromerobert/pdfarranger-dev/compare/master...gtk4). The main problem is that I do not see any reasonable way to have a code base compatible with both GTK3 and GTK4. As GTK3 is still our main target (because the main target of all Linux distribution), you'll have to wait a bit.

**Comment by K4LCIFER** [comment]
I await GTK4/libadwaita, as well!

**Comment by nekohayo** [comment]
It might be worth revisiting this, as even conservative distros like Debian 12 (stable) ship GTK4 applications by default now (and _lots_ of bugfixes have happened throughout 2023). At this point I see no reason to keep shipping GTK3 if you are able to migrate your codebase to GTK4. A new Ubuntu LTS version is coming out this month as well, and distributing as a Flatpak pretty much isolates you from "some distro doesn't offer GTK4" problems.

**Reply by dreua** [reply]
With Debian 11 going EOL in 3 month you have a point I think. As long as Gtk3 works and is not causing any issues other than some people thinking 4 would look nicer I personally don't feel like spending time on a migration.

**Comment by kbengs** [comment]
@linkmauve  My Gtk 4 Gridview wip mess can be found here: :smiley: 
https://github.com/kbengs/pdfarranger/tree/gtk4

Only way to open PDF is by command line argument.

**Comment by iaacornus** [comment]
hey i can help porting the project to gtk4 if you need help. but firstly, what's the current consensus on this

**Reply by kbengs** [reply]
I just created a gtk4 branch. We could try to do the work there. We can later move the commits to main branch when we think it's ready.

@linkmauve has done some work on gtk4 but I don't know the status. Maybe you could list what you have done? (and maybe what still need to be done)

Porting of Iconview I think is the hardest part. I have done some initial work on swapping Iconview with GridView. But I think it will take time before we are ready to do the swap. In the mean time we need to have Iconview working in Gtk4. I have rendering of thumbnails working in Gtk4+Iconview but nothing more.

**Reply by linkmauve** [reply]
Hi, I’ve just pushed my gtk4 branch here: https://github.com/linkmauve/pdfarranger/commits/gtk4

I’ve been porting the dialogs one by one, after the initial 3to4 commit and then manual changes to make things start at all. I’ve left out the iconview changes, since you already worked on that in your own branch.

I think after some small clean up my branch would be a good clean start.

**Reply by kbengs** [reply]
Is there any way to have a dialog moveable? I am thinking of for example the crop dialog where there is the "Apply" button. I would like to be able to move the dialog to the side so one can see what happens in main window.

The iconview part is here:
```
diff --git a/pdfarranger/iconview.py b/pdfarranger/iconview.py
index f1bdf43..e3d780d 100644
--- a/pdfarranger/iconview.py
+++ b/pdfarranger/iconview.py
@@ -14,8 +14,7 @@
 # with this program; if not, write to the Free Software Foundation, Inc.,
 # 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 
-from gi.repository import Gtk
-from gi.repository import Gdk
+from gi.repository import Gtk, Gdk, Graphene
 from math import pi
 
 
@@ -43,9 +42,11 @@ class CellRendererImage(Gtk.CellRenderer):
 
         return w0, h0, w1, h1, w2, h2, rotation
 
-    def do_render(self, window, _widget, _background_area, cell_area, _expose_area):
+    def do_snapshot(self, snapshot, _widget, _background_area, cell_area, _expose_area):
         if not self.page.thumbnail:
             return
+        rect = Graphene.Rect().init(cell_area.x, cell_area.y, cell_area.width, cell_area.height)
+        window = snapshot.append_cairo(rect)
 
         w0, h0, w1, h1, w2, h2, rotation = self.get_geometry()
         th = int(2 * self.th1 + self.th2)
@@ -86,20 +87,11 @@ class CellRendererImage(Gtk.CellRenderer):
         window.set_source_surface(self.page.thumbnail)
         window.paint()
 
-    def do_get_size(self, _widget, cell_area=None):
-        x = y = 0
+    def do_get_preferred_height_for_width(self, _iconview, height):
         th = int(2 * self.th1 + self.th2)
-        w = self.page.width_in_pixel() + th
         h = self.page.height_in_pixel() + th
-
-        if cell_area and w > 0 and h > 0:
-            x = self.get_property('xalign') * (
-                    cell_area.width - w - self.get_property('xpad'))
-            y = self.get_property('yalign') * (
-                    cell_area.height - h - self.get_property('ypad'))
-        w += 2 * self.get_property('xpad')
         h += 2 * self.get_property('ypad')
-        return int(x), int(y), w, h
+        return h, h
```
If anyone want to continue, just go ahead.

**Reply by linkmauve** [reply]
> Is there any way to have a dialog moveable? I am thinking of for example the crop dialog where there is the "Apply" button. I would like to be able to move the dialog to the side so one can see what happens in main window.

Not with the libadwaita dialogs I believe, nor on the GNOME platform at all. They decided to draw the dialogs on top of the rest of the window, instead of having another window above and tied to the parent one. We could decide to convert the dialogs where it’s useful to se behind into sidebars for instance. Edit: For now I’ve attempted to keep the UI/UX the closest to the existing gtk3 version, but some design changes would likely be welcome to improve things (or at least to not decrease usability). I’m thinking for instance about using a GtkPopoverMenu instead of dialogs for some things like splitting pages.

And thanks for the patch, I’ve applied it to my branch, it makes some thumbnails blurry though, but that’s good enough for a first attempt.

Would you prefer me to continue working on my gtk4 branch, or to periodically make a pull request to yours?

**Reply by kbengs** [reply]
I'd say you can keep the commits in your branch until you think they are ready for PR to https://github.com/pdfarranger/pdfarranger/tree/gtk4 I guess @iaacornus need to wait and see how things develop.

**Reply by vixalien** [reply]
I am also interested in contributing, if I could be filled in on the progress and where to look.
